src/System.Diagnostics.Process/src/System/Diagnostics/Process.Win32.cs

```diff
@@ -0,0 +1,168 @@
+                        shellExecuteInfo.nShow = Interop.Shell32.SW_SHOWNORMAL;
+                        break;
+                }
+
+                ShellExecuteHelper executeHelper = new ShellExecuteHelper(&shellExecuteInfo);
+                if (!executeHelper.ShellExecuteOnSTAThread())
+                {
+                    int error = executeHelper.ErrorCode;
+                    if (error == 0)
+                    {
+                        error = GetShellError(shellExecuteInfo.hInstApp);
+                    }
+
+                    switch (error)
+                    {
+                        case Interop.Errors.ERROR_BAD_EXE_FORMAT:
+                        case Interop.Errors.ERROR_EXE_MACHINE_TYPE_MISMATCH:
+                            throw new Win32Exception(error, SR.InvalidApplication);
+                        default:
+                            throw new Win32Exception(error);
+                    }
+                }
+
+                if (shellExecuteInfo.hProcess != IntPtr.Zero)
+                {
+                    SetProcessHandle(new SafeProcessHandle(shellExecuteInfo.hProcess));
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private int GetShellError(IntPtr error)
+        {
+            switch ((long)error)
+            {
+                case Interop.Shell32.SE_ERR_FNF:
+                    return Interop.Errors.ERROR_FILE_NOT_FOUND;
+                case Interop.Shell32.SE_ERR_PNF:
+                    return Interop.Errors.ERROR_PATH_NOT_FOUND;
+                case Interop.Shell32.SE_ERR_ACCESSDENIED:
+                    return Interop.Errors.ERROR_ACCESS_DENIED;
+                case Interop.Shell32.SE_ERR_OOM:
+                    return Interop.Errors.ERROR_NOT_ENOUGH_MEMORY;
+                case Interop.Shell32.SE_ERR_DDEFAIL:
+                case Interop.Shell32.SE_ERR_DDEBUSY:
+                case Interop.Shell32.SE_ERR_DDETIMEOUT:
+                    return Interop.Errors.ERROR_DDE_FAIL;
+                case Interop.Shell32.SE_ERR_SHARE:
+                    return Interop.Errors.ERROR_SHARING_VIOLATION;
+                case Interop.Shell32.SE_ERR_NOASSOC:
+                    return Interop.Errors.ERROR_NO_ASSOCIATION;
+                case Interop.Shell32.SE_ERR_DLLNOTFOUND:
+                    return Interop.Errors.ERROR_DLL_NOT_FOUND;
+                default:
+                    return (int)(long)error;
+            }
+        }
+
+        internal unsafe class ShellExecuteHelper
+        {
+            private Interop.Shell32.SHELLEXECUTEINFO* _executeInfo;
+            private bool _succeeded;
+
+            public ShellExecuteHelper(Interop.Shell32.SHELLEXECUTEINFO* executeInfo)
+            {
+                _executeInfo = executeInfo;
+            }
+
+            public void ShellExecuteFunction()
+            {
+                if (!(_succeeded = Interop.Shell32.ShellExecuteExW(_executeInfo)))
```

Comment:
Having said that it should not be possible to get there on Nano sinc eyou should make sure that setting UseShellExecute=true throws PNSE in that case. Although - chicken and egg - how do you know the API isn't available there?

We should have a tes tfor nano behavior please.

I have a way to ru nstuff on Nano - it's a HOWTO on the OneNote - or you could just wing it and run Nano in Jenkins (it's supported as an on demand job now).